kong/plugin_service.go

```diff
@@ -29,14 +31,58 @@ type AbstractPluginService interface {
 	ListAllForRoute(ctx context.Context, routeID *string) ([]*Plugin, error)
 	// Validate validates a Plugin against its schema
 	Validate(ctx context.Context, plugin *Plugin) (bool, string, error)
-	// GetSchema retrieves the schema of a plugin
+	// GetSchema retrieves the config schema of a plugin
 	GetSchema(ctx context.Context, pluginName *string) (map[string]interface{}, error)
+	// GetFullSchema retrieves the full schema of a plugin
+	GetFullSchema(ctx context.Context, pluginName *string) (map[string]interface{}, error)
+	// FillDefaults ingests plugin's defaults from its schema
+	FillDefaults(ctx context.Context, plugin *Plugin, schema map[string]interface{}) (*Plugin, error)
 }
 
 // PluginService handles Plugins in Kong.
 type PluginService service
 
-// GetSchema retrieves the schema of a plugin
+// GetFullSchema retrieves the full schema of a plugin
+func (s *PluginService) GetFullSchema(ctx context.Context,
+	pluginName *string) (map[string]interface{}, error) {
+	if isEmptyString(pluginName) {
+		return nil, fmt.Errorf("pluginName cannot be empty")
+	}
+	endpoint := fmt.Sprintf("/schemas/plugins/%v", *pluginName)
+	req, err := s.client.NewRequest("GET", endpoint, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	var schema map[string]interface{}
+	_, err = s.client.Do(ctx, req, &schema)
+	if err != nil {
+		return nil, err
+	}
+	return schema, nil
+}
+
+// FillDefaults ingests plugin's defaults from its schema
+func (s *PluginService) FillDefaults(ctx context.Context,
+	plugin *Plugin, schema map[string]interface{}) (*Plugin, error) {
+	jsonb, err := json.Marshal(&schema)
+	if err != nil {
+		return nil, err
+	}
+	gjsonSchema := gjson.ParseBytes((jsonb))
+	if plugin.Config == nil {
+		plugin.Config = make(Configuration)
+	}
+	plugin.Config = fillConfigRecord(gjsonSchema, plugin.Config)
+	if plugin.Protocols == nil {
+		plugin.Protocols = getDefaultProtocols(gjsonSchema)
+	}
+	if plugin.Enabled == nil {
+		plugin.Enabled = Bool(true)
+	}
+	return plugin, nil
+}
+
+// GetSchema retrieves the config schema of a plugin
```

Comment:
Mark this method as deprecated

Reply:
Is the underlying endpoint going away? It's a bit confusing that the gateway has two similar, but slightly different endpoints for this, but if they're going to keep both I suppose we should as well, unless the richer endpoint returns a response that's compatible with any existing usage of the config-only endpoint, in which case we should just keep a single function and switch it to the richer endpoint.

Reply:
The underlying endpoint was supposed to go away when the new one was introduced.
It couldn't happen because a lot of reliance was on the old endpoint - something that we should change by doing what is suggested here.

> the richer endpoint returns a response that's compatible with any existing usage of the config-only endpoint

The new richer output although has what the old endpoint had to offer, it is different. For this reason, it would be best to have a new function and deprecate the old one. Users of this library then understand a lot more and can take appropriate precautions.

Reply:
Okay, makes sense to me, marked as deprecated